### PR TITLE
TPMLib and Measured boot updates

### PR DESCRIPTION
**Pull request by lsubash**
- Added GetHashToExtend to bootloader common API to
  retrive digest to extend
- Hash calculations for stage component and config date updated
  to use bootloader common API
- Added functionality to extend KeyHashManifest digest

Signed-off-by: Subash Lakkimsetti <subash.lakkimsetti@intel.com>

### CODE REVIEW

File: BootloaderCorePkg/Stage2/Stage2.c [hunk 1]
```diff
@@ -13,26 +13,45 @@
...
+                               CbInfo->HashData, EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+
+      } else {
+        // Get Hash to extend based on component type and component src addresss
+        Status = GetHashToExtend ((UINT8) CbInfo->ComponentType,
+                                    GetCryptoHashAlg(PcdGet32(PcdMeasuredBootHashMask)),
+                                    CbInfo->CompBuf, CbInfo->CompLen, DigestHash);
+        if (Status == EFI_SUCCESS){
+          TpmExtendPcrAndLogEvent ( 0, (TPMI_ALG_HASH) GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask)), DigestHash,
+                                EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
```
Comment (myguitar) [inline BootloaderCorePkg/Stage2/Stage2.c]: Same comment as previous - make core more readable and re-usable.

File: BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c [hunk 2]
```diff
@@ -714,3 +715,78 @@ MatchHashInStore (
...
+  // Get componenet hash from hash store based on Componen Id and return if hash is valid
+  // Incase component hash is not avilable calculate hash from src buf and HashType provided.
+
+  // Get componenet hash from hash store based on Componen Id
+  if ((ComponentType >= COMP_TYPE_STAGE_1B) && (ComponentType < COMP_TYPE_INVALID)) {
+    Status = GetComponentHash (ComponentType, &Digest, &CompHashAlg);
+    if((Status == EFI_SUCCESS) && (CompHashAlg == HashType)) {
+
+      CompHashValid = TRUE;
+      CopyMem (HashData, Digest, DigestSize);
```
[the remark this answers is not in the document]
Comment (lsubash) [reply BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c]: GetcomponentHash() would return pointer address corresponding Component type.
GetHashToExtend() would accept an memory location and hash retrieved from GetcomponentHash() would be copied.

File: BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c [hunk 3]
```diff
@@ -714,3 +715,78 @@ MatchHashInStore (
...
+      return RETURN_SUCCESS;
+    }
+  }
+
+  if (CompHashValid == FALSE) {
```
[the remark this answers is not in the document]
Comment (lsubash) [reply BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c]: Yes, this condition is not mandatory, I will update to remove condition check, 
This code would be entered only if hash is not valid in above case.

File: BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c [hunk 4]
```diff
@@ -714,3 +715,78 @@ MatchHashInStore (
...
+    if (Src != NULL) {
+      // Calculate hash
+      DEBUG ((DEBUG_INFO, "Calculate Hash for  component as its not available in Component hash table \n"));
+      Status = CalculateHash ((UINT8 *)Src, Length, HashType, (UINT8 *) &DigestHash);
+      CopyMem (HashData, DigestHash, DigestSize);
```
[the remark this answers is not in the document]
Comment (lsubash) [reply BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c]: Thanks. I can remove copymem here. We can pass HashData ptr to CalculateHash.

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 5]
```diff
@@ -18,15 +21,32 @@ ExtendStage (
...
+
+      } else {
+        // Get Hash to extend based on component type and component src addresss
+        Status = GetHashToExtend ((UINT8) CbInfo->ComponentType,
+                                    GetCryptoHashAlg(PcdGet32(PcdMeasuredBootHashMask)),
+                                    CbInfo->CompBuf, CbInfo->CompLen, DigestHash);
+        if (Status == EFI_SUCCESS){
+          TpmExtendPcrAndLogEvent (0, (TPMI_ALG_HASH) GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask)), DigestHash ,
+                                EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+        }
```
[the remark this answers is not in the document]
Comment (lsubash) [reply BootloaderCorePkg/Stage1B/Stage1B.c]: HashData/DigestHash  passed in both cases are different.
So, TpmExtendPcrAndLogEvent is used separately. 
I will improve this code.

**Comment by lsubash** [comment]
Updated patch fixing comment provided by Aiden.

File: BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c [hunk 6]
```diff
@@ -714,3 +715,72 @@ MatchHashInStore (
...
+  IN       UINT32           Length,
+  OUT      UINT8           *HashData
+  )
+{
+  RETURN_STATUS        Status;
+  HASH_ALG_TYPE        CompHashAlg;
+  CONST UINT8         *Digest;
+  UINT8                DigestSize;
+
+  if(HashData == NULL) {
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c]: a space at if(

File: BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (myguitar) [reply BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c]: Got it. Thanks.

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 7]
```diff
@@ -7,26 +7,52 @@
 
 #include "Stage1B.h"
 
+
+
+
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1B/Stage1B.c]: Remove empty lines

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 8]
```diff
@@ -7,26 +7,52 @@
...
+      } else {
+        // Get Hash to extend based on component type and component src addresss
+        Status = GetHashToExtend ((UINT8) CbInfo->ComponentType,
+                                    MbHashType, CbInfo->CompBuf, CbInfo->CompLen, DigestHash);
+        if (Status == EFI_SUCCESS){
+          TpmExtendPcrAndLogEvent (0, MbTmpAlgHash, DigestHash,
+                                EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
+        }
+      }
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1B/Stage1B.c]: What about this?
UINT8 *HashPtr;

if ((CbInfo->HashAlg == MbHashType) && (CbInfo->HashData != NULL)) {
  HashPtr = CbInfo->HashData;
  Status = EFI_SUCCESS;
} else {
  // Get Hash to extend based on component type and component src addresss
  Status = GetHashToExtend ((UINT8) CbInfo->ComponentType, MbHashType, CbInfo->CompBuf, CbInfo->CompLen, DigestHash);								
  HashPtr = DigestHash;
}

if ((Status == EFI_SUCCESS) && (HashPtr != NULL)) {
  TpmExtendPcrAndLogEvent (0, MbTmpAlgHash, HashPtr, EV_POST_CODE, POST_CODE_STR_LEN, (UINT8 *)EV_POSTCODE_INFO_POST_CODE);
}

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 9]
```diff
@@ -181,13 +209,36 @@ AppendHashStore (
...
+                          NULL,
+                          Stage1bParam->KeyHashManifestHash);
   }
   if (EFI_ERROR (Status)) {
+    Stage1bParam->KeyHashManifestHashValid = 0;
     return EFI_SECURITY_VIOLATION;
   }
 
+  if (MEASURED_BOOT_ENABLED()) {
+    if (PcdGet8(PcdCompSignHashAlg) == GetCryptoHashAlg (PcdGet32(PcdMeasuredBootHashMask))) {
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1B/Stage1B.c]: UINT8 HashAlg;

HashAlg = GetCryptoHashAlg (PcdGet32(PcdMeasuredBootHashMask));
if ( == HashAlg) {
} else {
GetHashToExtend(,HashAlg,,,,);
}
}

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 10]
```diff
@@ -627,14 +692,26 @@ ContinueFunc (
...
+  if (MEASURED_BOOT_ENABLED()) {
     if (GetBootMode() != BOOT_ON_S3_RESUME) {
+        // Extend External Config Data hash
         if (Stage1bParam->ConfigDataHashValid == 1) {
-          TpmExtendConfigData (Stage1bParam->ConfigDataHash,
-            PcdGet8(PcdCompSignHashAlg),
-            (UINT8 *) Stage1bParam->CfgDataAddr,
-            ((CDATA_BLOB *) Stage1bParam->CfgDataAddr)->UsedLength);
+          TpmExtendPcrAndLogEvent ( 1,
+                    (TPMI_ALG_HASH) GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask)),
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1B/Stage1B.c]: Use a variable for GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask)).
if ConfigDataHashValid == 1 and KeyHashManifestHashValid == 1, the same function can be called twice.

File: BootloaderCorePkg/Stage2/Stage2.c [hunk 11]
```diff
@@ -13,26 +13,51 @@
...
+
+  //Convert Measured boot Hash Mask to TPMI_ALG_HASH (TPM ALG ID)
+  MbTmpAlgHash = (TPMI_ALG_HASH) GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask));
 
   //Check the boot mode
-   BootMode = GetBootMode();
+  BootMode = GetBootMode();
   if (MEASURED_BOOT_ENABLED() && (BootMode != BOOT_ON_S3_RESUME)) {
-    //Extend hash for Component Types  Payload, Epayload and FWPayload
+  //Extend  hash if ComponentType is payload
```
Comment (myguitar) [inline BootloaderCorePkg/Stage2/Stage2.c]: indent here.

File: BootloaderCorePkg/Stage2/Stage2.c [hunk 12]
```diff
@@ -13,26 +13,51 @@
...
     if ((CbInfo != NULL ) &&  ((CbInfo->ComponentType == COMP_TYPE_PAYLOAD)
                               || (CbInfo->ComponentType == FLASH_MAP_SIG_EPAYLOAD)
                               || (CbInfo->ComponentType == COMP_TYPE_PAYLOAD_FWU ))) {
-      TpmExtendStageHash((UINT8) CbInfo->ComponentType, CbInfo->HashData, CbInfo->HashAlg,
-                          CbInfo->CompBuf, CbInfo->CompLen);
+      // Check Hash data alg match to PcdMeasuredBootHashMask
+      if ((CbInfo->HashAlg == MbHashType) && (CbInfo->HashData != NULL)) {
+        // Extend CbInfo->HashData if hashalg is valid
+        TpmExtendPcrAndLogEvent (0, MbTmpAlgHash, CbInfo->HashData,
```
Comment (myguitar) [inline BootloaderCorePkg/Stage2/Stage2.c]: HashPtr = CbInfo->HashData or HashPtr = DigestHash.
TpmExtendPcrAndLogEvent (,,HashPtr,,,);

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 13]
```diff
@@ -258,12 +309,26 @@ CreateConfigDatabase (
         }
       }
       if (ExtCfgAddPtr != NULL) {
+       if (MEASURED_BOOT_ENABLED()) {
+         // Check if Stage1bParam->ConfigDataHash generated matches PcdMeasuredBootHashMask
+         if (PcdGet8(PcdCompSignHashAlg) == GetCryptoHashAlg (PcdGet32(PcdMeasuredBootHashMask))) {
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1B/Stage1B.c]: Same here - GetCryptoHashAlg (PcdGet32(PcdMeasuredBootHashMask)).

File: BootloaderCorePkg/Stage1B/Stage1B.c (hunk 8 above)
Comment (lsubash) [reply BootloaderCorePkg/Stage1B/Stage1B.c]: yes. We could do. Thanks!

File: BootloaderCorePkg/Stage1B/Stage1B.c (hunk 9 above)
Comment (lsubash) [reply BootloaderCorePkg/Stage1B/Stage1B.c]: Will Update,

File: BootloaderCorePkg/Stage1B/Stage1B.c (hunk 10 above)
Comment (lsubash) [reply BootloaderCorePkg/Stage1B/Stage1B.c]: Variable can be added for GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask)).
TpmExtendPcrAndLogEvent hash be added separately for config data and KeyHashManifestHash extend.

**Comment by lsubash** [comment]
Updated PR as per Aiden feedback.

File: BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c [hunk 14]
```diff
@@ -714,3 +715,72 @@ MatchHashInStore (
...
+      return RETURN_SUCCESS;
+    }
+  }
+
+  // Calculate hash for a ComponentType if hash is not retrieved from GetComponentHash
+  if ((Src != NULL) && (Length > 0)) {
+    DEBUG ((DEBUG_INFO, "Calculate Hash for component Type 0x%x as its not available in Component hash table \n", ComponentType));
+    Status = CalculateHash ((UINT8 *)Src, Length, HashType, (UINT8 *) HashData);
+  } else{
+    DEBUG ((DEBUG_INFO, "Valid Hash is not available!!\n"));
```
Comment (sagraw2) [inline BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c]: Remove this debug statement.

File: BootloaderCorePkg/Stage1B/Stage1B.c [hunk 15]
```diff
@@ -219,6 +276,7 @@ CreateConfigDatabase (
   SIGNATURE_HDR            *SignHdr;
   UINT32                    CfgDataBase;
   UINT32                    CfgDataLength;
+  HASH_ALG_TYPE        MbHashType;
```
Comment (sagraw2) [inline BootloaderCorePkg/Stage1B/Stage1B.c]: Is indentation correct here?

**Comment by lsubash** [comment]
Updated as per Sachin's comments.

**Pull request merged**